From 081a5b2c84e299b82c6d37da693b6e097555ad84 Mon Sep 17 00:00:00 2001
From: Kihwal Lee <kihwal@apache.org>
Date: Thu, 5 Nov 2015 09:27:36 -0600
Subject: [PATCH 1196/2748] HDFS-4937. ReplicationMonitor can infinite-loop in
 BlockPlacementPolicyDefault#chooseRandom().
 Contributed by Kihwal Lee. (cherry picked from
 commit ff47f35deed14ba6463cba76f0e6a6c15abb3eca)

(cherry picked from commit 616ed9084be4dc337c2ececa7aecc4ab7899a75a)

Change-Id: I7000e9182a6460a817df618aa41387050640ecba
---
 .../BlockPlacementPolicyDefault.java               |    9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockPlacementPolicyDefault.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockPlacementPolicyDefault.java
index 4858bb5..e03d5d2 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockPlacementPolicyDefault.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockPlacementPolicyDefault.java
@@ -621,6 +621,7 @@ protected DatanodeStorageInfo chooseRandom(int numOfReplicas,
       
     int numOfAvailableNodes = clusterMap.countNumOfAvailableNodes(
         scope, excludedNodes);
+    int refreshCounter = numOfAvailableNodes;
     StringBuilder builder = null;
     if (LOG.isDebugEnabled()) {
       builder = debugLoggingBuilder.get();
@@ -674,6 +675,14 @@ protected DatanodeStorageInfo chooseRandom(int numOfReplicas,
         // If no candidate storage was found on this DN then set badTarget.
         badTarget = (i == storages.length);
       }
+      // Refresh the node count. If the live node count became smaller,
+      // but it is not reflected in this loop, it may loop forever in case
+      // the replicas/rack cannot be satisfied.
+      if (--refreshCounter == 0) {
+        numOfAvailableNodes = clusterMap.countNumOfAvailableNodes(scope,
+            excludedNodes);
+        refreshCounter = numOfAvailableNodes;
+      }
     }
       
     if (numOfReplicas>0) {
-- 
1.7.9.5

